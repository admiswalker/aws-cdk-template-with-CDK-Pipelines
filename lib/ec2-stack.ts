import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';
import * as fs from 'fs';


interface Ec2StackProps extends StackProps {
  prj_name: string;
  vpc: ec2.Vpc;
  ssm_iam_role: iam.Role;
}
export class Ec2Stack extends Stack {
  constructor(scope: Construct, id: string, props: Ec2StackProps) {
    super(scope, id, props);

    const cloud_config = ec2.UserData.forLinux({shebang: ''})
    const user_data_script = fs.readFileSync('./lib/ec2-stack_user-data.yaml', 'utf8');
    cloud_config.addCommands(user_data_script)
    const multipartUserData = new ec2.MultipartUserData();
    multipartUserData.addPart(ec2.MultipartBody.fromUserData(cloud_config, 'text/cloud-config; charset="utf8"'));
    
    const ec2_instance = new ec2.Instance(this, props.prj_name+'-'+this.constructor.name+'-general_purpose_ec2', {
      instanceType: new ec2.InstanceType('t3.nano'), // 1 Core, 1 GB
//    machineImage: ec2.MachineImage.genericLinux({'us-west-2': 'ami-XXXXXXXXXXXXXXXXX'}),
      machineImage: new ec2.AmazonLinuxImage({
        generation: ec2.AmazonLinuxGeneration.AMAZON_LINUX,
        edition: ec2.AmazonLinuxEdition.STANDARD,
        virtualization: ec2.AmazonLinuxVirt.HVM,
        storage: ec2.AmazonLinuxStorage.GENERAL_PURPOSE,
      }),
      vpc: props.vpc,
//    blockDevices: [{
//	    deviceName: '/dev/sda1',
//	    volume: ec2.BlockDeviceVolume.ebs(30),
//    }],
      vpcSubnets: props.vpc.selectSubnets({
        subnetGroupName: 'Private',
      }),
      role: props.ssm_iam_role,
      userData: multipartUserData,
    });

  }
}
