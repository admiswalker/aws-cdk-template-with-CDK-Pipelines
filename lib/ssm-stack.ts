import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import * as iam from 'aws-cdk-lib/aws-iam';


interface SsmStackProps extends StackProps {
  prj_name: string;
  vpc: ec2.Vpc;
}
export class SsmStack extends Stack {
  public readonly iam_role: iam.Role;

  constructor(scope: Construct, id: string, props: SsmStackProps) {
    super(scope, id, props);

    this.iam_role = new iam.Role(this, props.prj_name+'-'+this.constructor.name+'-iam_role_for_ssm', {
      assumedBy: new iam.ServicePrincipal('ec2.amazonaws.com'),
      managedPolicies: [
        iam.ManagedPolicy.fromAwsManagedPolicyName('AmazonSSMManagedInstanceCore'),
        iam.ManagedPolicy.fromAwsManagedPolicyName('CloudWatchAgentAdminPolicy'),
      ],
    });

    props.vpc.addInterfaceEndpoint(props.prj_name+'-'+this.constructor.name+'-InterfaceEndpoint_ssm', {
      service: ec2.InterfaceVpcEndpointAwsService.SSM,
    });
    props.vpc.addInterfaceEndpoint(props.prj_name+'-'+this.constructor.name+'-InterfaceEndpoint_ec2_messages', {
      service: ec2.InterfaceVpcEndpointAwsService.EC2_MESSAGES,
    });
    props.vpc.addInterfaceEndpoint(props.prj_name+'-'+this.constructor.name+'-InterfaceEndpoint_ssm_messages', {
      service: ec2.InterfaceVpcEndpointAwsService.SSM_MESSAGES,
    });

  }
}
