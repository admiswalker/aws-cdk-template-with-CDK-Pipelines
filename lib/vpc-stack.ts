import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { InstanceType, NatInstanceImage, NatProvider } from 'aws-cdk-lib/aws-ec2';


interface VpcStackProps extends StackProps {
    prj_name: string;
}
export class VpcStack extends Stack {
  public readonly vpc: ec2.Vpc;

  constructor(scope: Construct, id: string, props: VpcStackProps) {
    super(scope, id, props);

    this.vpc = new ec2.Vpc(this, props.prj_name+'-'+this.constructor.name+'-vpc_for_ec2_and_ssm', {
      cidr: '10.0.0.0/16',
      natGateways: 1,
      natGatewayProvider: ec2.NatProvider.instance({
        instanceType: new InstanceType('t3.nano'),
        machineImage: new NatInstanceImage(),
      }),
      subnetConfiguration: [
        {
          name: 'Public',
          subnetType: ec2.SubnetType.PUBLIC,
          cidrMask: 27,
        },
        {
          name: 'Private',
          subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
          cidrMask: 27,
        },
      ],
    });

  }
}