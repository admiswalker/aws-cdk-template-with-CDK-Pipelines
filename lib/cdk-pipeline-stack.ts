import { Stack, StackProps } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import * as codecommit from 'aws-cdk-lib/aws-codecommit';
import {
    CodeBuildStep,
    CodePipeline,
    CodePipelineSource,
    ShellStep,
} from "aws-cdk-lib/pipelines";
import { CdkPipelineStageStack } from './cdk-pipeline-stage-stack';


interface CdkPipelineStackProps extends StackProps {
    prj_name: string;
}
export class CdkPipelineStack extends Stack {
  constructor(scope: Construct, id: string, props: CdkPipelineStackProps) {
    super(scope, id, props);

    // const repo = new codecommit.Repository(this, 'aws-cdk-template-with-CDK-Pipelines', { // Creates a new, empty repository
    //     repositoryName: 'aws-cdk-template-with-CDK-Pipelines',
    // });
    const repo = codecommit.Repository.fromRepositoryName(this, 'repo', 'aws-cdk-template-with-CDK-Pipelines'); // Use existing repository

    const pipeline = new CodePipeline(this, props.prj_name+'-'+this.constructor.name+'-aws-cdk-template-with-CDK-Pipelines',{
        pipelineName: props.prj_name+'-'+this.constructor.name+'-aws-cdk-template-with-CDK-Pipelines',
        synth: new CodeBuildStep('SynthStep', {
            input: CodePipelineSource.codeCommit(repo, 'main'),
            installCommands: [
                'npm install -g aws-cdk'
            ],
            commands: [
                'npm ci',
                'npm run build',
                'npx cdk synth'
            ],
        }),
    });

    const deploy_stack = new CdkPipelineStageStack(this, props.prj_name+'Deploy', {
        prj_name: props.prj_name,
        env: props.env,
    });
    const deployStage = pipeline.addStage(deploy_stack);
  }
}


// memo
// - AWS の CDK pipelines と aws-codepipeline の違い
//   https://zenn.dev/siroken3/scraps/769002c03b31d6
// - CDK Pipelineを導入するときの手順メモ
//   https://zenn.dev/pirosikick/scraps/a3d843cc54fcbf
// - class CodeCommitSourceAction
//   https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-codepipeline-actions.CodeCommitSourceAction.html
// - CREATE NEW PIPELINE - TypeScript Workshop
//   https://cdkworkshop.com/20-typescript/70-advanced-topics/200-pipelines/3000-new-pipeline.html
