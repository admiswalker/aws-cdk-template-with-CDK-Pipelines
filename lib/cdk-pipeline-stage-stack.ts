import { Construct } from 'constructs';
import { Stage, StageProps } from 'aws-cdk-lib';
import { AwsCdkTemplateStack } from '../lib/aws-cdk-template-stack';
import { VpcStack } from '../lib/vpc-stack';
import { SsmStack } from '../lib/ssm-stack';
import { Ec2Stack } from '../lib/ec2-stack';


interface CdkPipelineStageStackProps extends StageProps {
    prj_name: string;
}
export class CdkPipelineStageStack extends Stage {
  constructor(scope: Construct, id: string, props: CdkPipelineStageStackProps) {
    super(scope, id, props);

    const aws_cdk_template_stack = new AwsCdkTemplateStack(this, 'AwsCdkTemplateStack', {
        env: props.env,
      });
      
      const vpc_stack = new VpcStack(this, props.prj_name+'-VpcStack', {
        prj_name: props.prj_name,
        env: props.env,
      });
      
      const ssm_stack = new SsmStack(this, props.prj_name+'-SsmStack', {
        prj_name: props.prj_name,
        env: props.env,
        vpc: vpc_stack.vpc,
      });
      
      const ec2_stack = new Ec2Stack(this, props.prj_name+'-Ec2Stack', {
        prj_name: props.prj_name,
        env: props.env,
        vpc: vpc_stack.vpc,
        ssm_iam_role: ssm_stack.iam_role,
      });

  }
}

// memo
// - ADD APPLICATION TO PIPELINE - TypeScript Workshop
//   https://cdkworkshop.com/20-typescript/70-advanced-topics/200-pipelines/4000-build-stage.html
