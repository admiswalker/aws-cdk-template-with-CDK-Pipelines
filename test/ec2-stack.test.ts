import * as cdk from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { VpcStack } from '../lib/vpc-stack';
import { SsmStack } from '../lib/ssm-stack';
import { Ec2Stack } from '../lib/ec2-stack';


test('Ec2Stack', () => {
  
  const app = new cdk.App();

  const prj_name = 'AwsCdkTemplate';
  const env = {
    account: 'test-account',
    region: 'test-region'
  }

  const vpc_stack = new VpcStack(app, prj_name+'-VpcStack', {
    prj_name: prj_name,
    env: env,
  });
  
  const ssm_stack = new SsmStack(app, prj_name+'-SsmStack', {
    prj_name: prj_name,
    env: env,
    vpc: vpc_stack.vpc,
  });
  
  // WHEN
  const ec2_stack = new Ec2Stack(app, prj_name+'-Ec2Stack', {
    prj_name: prj_name,
    env: env,
    vpc: vpc_stack.vpc,
    ssm_iam_role: ssm_stack.iam_role,
  });
  
  // THEN
  expect(Template.fromStack(ec2_stack).toJSON()).toMatchSnapshot();

});
