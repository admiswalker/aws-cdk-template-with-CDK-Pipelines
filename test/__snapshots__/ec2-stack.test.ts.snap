// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ec2Stack 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter": Object {
      "Default": "/aws/service/ami-amazon-linux-latest/amzn-ami-hvm-x86_64-gp2",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::Image::Id>",
    },
  },
  "Resources": Object {
    "AwsCdkTemplateEc2Stackgeneralpurposeec221780F40": Object {
      "Properties": Object {
        "AvailabilityZone": "dummy1a",
        "IamInstanceProfile": Object {
          "Ref": "AwsCdkTemplateEc2Stackgeneralpurposeec2InstanceProfile3410BCD6",
        },
        "ImageId": Object {
          "Ref": "SsmParameterValueawsserviceamiamazonlinuxlatestamznamihvmx8664gp2C96584B6F00A464EAD1953AFF4B05118Parameter",
        },
        "InstanceType": "t3.nano",
        "SecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "AwsCdkTemplateEc2Stackgeneralpurposeec2InstanceSecurityGroup50111657",
              "GroupId",
            ],
          },
        ],
        "SubnetId": Object {
          "Fn::ImportValue": "AwsCdkTemplate-VpcStack:ExportsOutputRefAwsCdkTemplateVpcStackvpcforec2andssmPrivateSubnet1Subnet36719CEC6B0F3C7C",
        },
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "AwsCdkTemplate-Ec2Stack/AwsCdkTemplate-Ec2Stack-general_purpose_ec2",
          },
        ],
        "UserData": Object {
          "Fn::Base64": Object {
            "Fn::Join": Array [
              "",
              Array [
                "Content-Type: multipart/mixed; boundary=\\"+AWS+CDK+User+Data+Separator==\\"
MIME-Version: 1.0

--+AWS+CDK+User+Data+Separator==
Content-Type: text/cloud-config; charset=\\"utf8\\"
Content-Transfer-Encoding: base64

",
                Object {
                  "Fn::Base64": "
#cloud-config
cloud_final_modules:
  - [users-groups, always]
users:
  - default
  - name: admis
    groups: [wheel]
    sudo: [\\"ALL=(ALL) NOPASSWD:ALL\\"]
    shell: /bin/bash
    ssh-authorized-keys:
      - ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIGsHMbqflou/Pow660rnIJq8wdtCnoJhzZfh6BAOWrxQ admis@admis-MS-7A34

",
                },
                "
--+AWS+CDK+User+Data+Separator==--
",
              ],
            ],
          },
        },
      },
      "Type": "AWS::EC2::Instance",
    },
    "AwsCdkTemplateEc2Stackgeneralpurposeec2InstanceProfile3410BCD6": Object {
      "Properties": Object {
        "Roles": Array [
          Object {
            "Fn::ImportValue": "AwsCdkTemplate-SsmStack:ExportsOutputRefAwsCdkTemplateSsmStackiamroleforssmF524EAAE214DB9E2",
          },
        ],
      },
      "Type": "AWS::IAM::InstanceProfile",
    },
    "AwsCdkTemplateEc2Stackgeneralpurposeec2InstanceSecurityGroup50111657": Object {
      "Properties": Object {
        "GroupDescription": "AwsCdkTemplate-Ec2Stack/AwsCdkTemplate-Ec2Stack-general_purpose_ec2/InstanceSecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "Name",
            "Value": "AwsCdkTemplate-Ec2Stack/AwsCdkTemplate-Ec2Stack-general_purpose_ec2",
          },
        ],
        "VpcId": Object {
          "Fn::ImportValue": "AwsCdkTemplate-VpcStack:ExportsOutputRefAwsCdkTemplateVpcStackvpcforec2andssm1E664AD5227AD8D4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
